core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncRetryableStage.java

```diff
@@ -122,65 +112,75 @@ private RetryExecutor(SdkHttpFullRequest request, RequestExecutionContext contex
-                    SdkException retryableException = handleSdkException(
-                        Response.fromFailure(throwable, null));
-                    retryHandler.setLastRetriedException(retryableException);
-                    // Notify the response handler on each retry. Note that this does not notify
-                    // on the last attempt by design. This is done in the generated client code so that
-                    // any exceptions that happen before we get to the retryable stage are also delivered to the
-                    // response handler
-                    deliverExceptionToResponseHandler(retryableException);
+                    maybeRetryError(err, future);
+                }
+            } catch (Throwable t) {
+                future.completeExceptionally(t);
+            }
+        }
+
+        private void maybeRetryResponse(Response<OutputT> resp, CompletableFuture<Response<OutputT>> future) {
+            if (resp.isSuccess()) {
+                retryHandler.releaseRetryCapacity();
+                future.complete(resp);
+            } else {
+                SdkException err = resp.exception();
+
+                if (RetryUtils.isClockSkewException(err)) {
+                    int clockSkew = ClockSkewUtil.parseClockSkewOffset(resp.httpResponse());
+                    dependencies.updateTimeOffset(clockSkew);
+                }
+
+                if (shouldRetry(resp.httpResponse(), resp.exception())) {
+                    retryHandler.setLastRetriedException(err);
                     executeRetry(future);
+                } else {
+                    future.completeExceptionally(err);
                 }
-            } catch (Exception e) {
-                future.completeExceptionally(e);
             }
-            return null;
         }
 
-        /**
-         * Notify the response handler on each retry. Note that this does not notify on the last attempt by design. This is done
-         * in the generated client code so that any exceptions that happen before we get to the retryable stage are also
-         * delivered to the response handler.
-         */
-        private void deliverExceptionToResponseHandler(SdkException retryableException) {
-            responseHandler.exceptionOccurred(retryableException);
+        private void maybeRetryError(Throwable err, CompletableFuture<Response<OutputT>> future) {
+            if (err instanceof CompletionException) {
+                err = err.getCause();
+            }
+
+            if (err instanceof NonRetryableException) {
+                future.completeExceptionally(err);
+                return;
+            }
+
+            // TODO(dongie): We need to wrap into SdkException in order to call setLastRetriedException. Is this necessary?
+            // Can it just take a Throwable?
+            SdkException sdkException = err instanceof SdkException ?
```

Comment:
TODO?

Reply:
Not sure what you mean. Do you mean to remove the "(dongie)" part?